src/allocator_defrag.c

```diff
@@ -0,0 +1,478 @@
+/* Get the bin index in bin array from the reg_size.
+ *
+ * these are reverse engineered mapping of reg_size -> binind. We need this information because the utilization query
+ * returns the size of the buffer and not the bin index, and we need the bin index to access it's usage information
+ *
+ * Note: In case future PR will return the binind (that is better API anyway) we can get rid of
+ * these conversion functions*/
+inline unsigned jemalloc_sz2binind_lgq3(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 24, 32, 40, 48, 56, 64
+        return (sz >> 3) - 1;
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return get_binind_normal(sz, LG_QUANTOM_OFFSET_3, last_sz_in_group_pow2);
+}
+
+inline unsigned jemalloc_sz2binind_lgq4(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 32, 48, 64
+        return (sz >> 4);
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return get_binind_normal(sz, LG_QUANTOM_OFFSET_4, last_sz_in_group_pow2);
+}
+
+/* -----------------------------------------------------------------------------
+ * Get INFO string about the defrag.
+ * -------------------------------------------------------------------------- */
+/*
+ * add defrag info string into info
+ */
+sds defrag_jemalloc_get_fragmentation_info(sds info) {
+    if (!defrag_supported) return info;
+    je_binfo *binfo;
+    je_busage *busage;
+    unsigned nbins = arena_bin_conf.nbins;
+    if (nbins > 0) {
+        info = sdscatprintf(info,
+                            "jemalloc_quantom:%d\r\n"
+                            "hit_ratio:%lu%%,hits:%lu,misses:%lu\r\n"
+                            "hit_bytes:%lu,miss_bytes:%lu\r\n"
+                            "ncalls_util_batches:%lu,ncalls_util_ptrs:%lu\r\n",
+                            (int)jemalloc_quantom,
+                            (usage_latest.stats.hits + usage_latest.stats.misses)
+                                ? usage_latest.stats.hits / (usage_latest.stats.hits + usage_latest.stats.misses)
+                                : 0,
+                            usage_latest.stats.hits, usage_latest.stats.misses, usage_latest.stats.hit_bytes,
+                            usage_latest.stats.miss_bytes, usage_latest.stats.ncalls, usage_latest.stats.nptrs);
+        for (unsigned j = 0; j < nbins; j++) {
+            binfo = &arena_bin_conf.bin_info[j];
+            busage = &usage_latest.bins_usage[j];
+            info = sdscatprintf(info,
+                                "[%d][%lu]::"
+                                "nregs:%lu,nslabs:%lu,nnonfull:%lu,"
+                                "hit_rate:%lu%%,hit:%lu,miss:%lu,nmalloc:%lu,ndealloc:%lu\r\n",
```

Comment:
This is even weirder syntax. I don't think all this low level defrag information should be printed by default. Also, do we expect anyone to really be able to use this information? Exposing data is a one way door.

Reply:
This information is at the bin level, we don't have the equivalent today and it is useful for someone debugging defrag or if they want to have advance enable\disable defrag logic. I agree it's a lot of very low level details, if we remove it from `info all` and only output if `info defrag` is queried, would this make more sense? or do you think we should use other mechanism to expose this info in case of need?

Reply:
@zuiderkwast @madolson would appreciate your thoughts on the `info defrag` section. If you agree it is useful, how would you suggest to combine each bin's info? I don't want to have a line for each stat member of each bin...